### ISSUE TITLE
created_by_id is not being populated on save

### ISSUE THREAD

**Issue by dleard** [bcgov/cas-reporting#416]
#### Description:

As a developer, I want the `created_by_id` and `updated_by_id` fields to be populated automatically using a Postgres trigger.

Note: if starting this work before #459, the extension needs to be installed in a first small PR, so the work is not blocked

#### Acceptance Criteria:

Given I am a developer,
When I create or update a data model that inherits from `TimeStampedModel`,
Then the `created_by_id` and `updated_by_id` fields should be populated automatically.

#### Development Checklist:

- [x] Follow [django-pgtrigger installation](https://django-pgtrigger.readthedocs.io/en/4.13.3/installation/)
- [x] Use the middleware from RLS exploration(https://github.com/bcgov/cas-registration/pull/2558/files#diff-b11e4335e5b8acce6d932f0d5af23347da98356429c9629b034ffb46402aff0a) to set the user_guid on each DB connection
- [x] Add a trigger to `TimeStampedModel` data model to populate `created_by_id` and `updated_by_id` based on Insert and Update operations and the user associated to the DB connection
- [x] Update backend tests to make sure it is working as expected
- [x] Remove existing references to `set_create_or_update`

#### Definition of Ready (Note: If any of these points are not applicable, mark N/A)

- [ ] User story is included
- [ ] User role and type are identified
- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if required)
- [ ] Design / Solution is accepted by Product Owner
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

#### Definition of Done (Note: If any of these points are not applicable, mark N/A)

- [ ] Acceptance criteria are tested by the CI pipeline
- [ ] UI meets accessibility requirements
- [ ] Configuration changes are documented, documentation and designs are updated
- [ ] Passes code peer-review
- [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
- [ ] Ticket is ready to be merged to main branch
- [ ] Can be demoed in Sprint Review
- [ ] Bugs or future work cards are identified and created
- [ ] Reviewed and approved by Product Owner

#### Notes:

-

**Comment by dleard** [comment]
@Sepehr-Sobhani Tagging you because you have the most experience with these fields. Not sure if this was a known bug already?

**Comment by Sepehr-Sobhani** [comment]
> [@Sepehr-Sobhani](https://github.com/Sepehr-Sobhani) Tagging you because you have the most experience with these fields. Not sure if this was a known bug already?

This is not a bug but a known issue we have experienced since the beginning. We have created explicit functions to set the fields `created_by`, `updated_by`, and `archived_by`, which you can find here: https://github.com/bcgov/cas-registration/blob/develop/docs/backend/data-models.md#audit-trail-implementation

**Comment by dleard** [comment]
Thanks @Sepehr-Sobhani ! I knew you were the right guy to ask :)
I suppose I should turn this "bug" card into a tech debt to make sure that the endpoints creating data in the timestamped tables we've created in reporting are calling those functions.

**Comment by pbastia** [comment]
@Sepehr-Sobhani @dleard  I wonder if we could enforce in our abstract `TimeStampedModel` that `created_by` is not null. This will force devs to pass the user upon each record creation and greatly reduce the risk of forgetting to 1) call the `set_create_or_update()` function and 2) write a test that the user is properly setup

**Comment by Sepehr-Sobhani** [comment]
> [@Sepehr-Sobhani](https://github.com/Sepehr-Sobhani) [@dleard](https://github.com/dleard) I wonder if we could enforce in our abstract `TimeStampedModel` that `created_by` is not null. This will force devs to pass the user upon each record creation and greatly reduce the risk of forgetting to 1) call the `set_create_or_update()` function and 2) write a test that the user is properly setup

Then we won't be able to call `save` method if we are not passing the user. One possible solution would be to pass the user directly to the `save` method.

**Comment by Sepehr-Sobhani** [comment]
@pbastia @dleard What if we set the user_guid as the postgres session in a middleware, something like:

```python
from django.db import connection

class UserUUIDMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        try:
            user_uuid = request.headers.get("...")
            if user_uuid:
                with connection.cursor() as cursor:
                    cursor.execute("<SET SESSION HERE> = %s;", [user_uuid])
            response = self.get_response(request)
        finally:
            with connection.cursor() as cursor:
                cursor.execute("<RESET SESSION HERE>;")
        return response
```

And then having a trigger in Postgres to set audit fields for us.

**Comment by pbastia** [comment]
Yes @Sepehr-Sobhani ! We've been talking about doing something like that. It does require a bunch of non-django SQL code to maintain though.
But we'll be implementing that middleware for this card anyways: https://github.com/bcgov/cas-reporting/issues/365

**Comment by pbastia** [comment]
@dleard should we just not worry about this one until we have RLS / automatic record amendment setup? Or do you reckon this would be something separate, at least initially?

**Comment by dleard** [comment]
It could be related if we decide to use postgres triggers to do this for us. We can use the `my.guid` setting that gets set via middleware in the RLS proof of concept. If we want to keep this a pure django thing then it's probably entirely separate

**Comment by dleard** [comment]
I'll add the attn_developer tag. I think it's worth some discussion about how to proceed here

**Comment by pbastia** [comment]
@Sepehr-Sobhani suggested something like https://django-pgtrigger.readthedocs.io/en/4.13.3/ could be used

**Comment by Sepehr-Sobhani** [comment]
@pbastia / @dleard Ticket is refined.

**Comment by Sepehr-Sobhani** [comment]
This ticket is unblocked by merging PR: https://github.com/bcgov/cas-registration/pull/2729
